File: scripts/unet.py

# https://github.com/mateuszbuda/brain-segmentation-pytorch/

import torch
import torch.nn as nn
import torch.nn.functional as F
from torch.utils.data import Dataset

import numpy as np
import os
import copy
import random
import torchvision
from torchvision import datasets, models, transforms

from skimage.io import imread

exec(open('scripts/brainseg_utils.py').read())
exec(open('scripts/brainseg_transforms.py').read())

# ------------------------------------------------------------------------------

train_dir = "data/kaggle_3m_train"
valid_dir = "data/kaggle_3m_valid"


# ------------------- load data ------------------------------------------------

image_slices = []
mask_slices = []
filepath = "data/kaggle_3m_train/TCGA_CS_4942_19970222/TCGA_CS_4942_19970222_1_mask.tif"
image_slices.append(imread(filepath, as_gray=True))
image_slices[0].shape
len(image_slices)
volumes = {}
patient_id = 2222
volumes[patient_id] = np.array(image_slices[1:-1])
len(image_slices[1:-1])
volumes[patient_id].shape
            
class BrainSegmentationDataset(Dataset):
    """Brain MRI dataset for FLAIR abnormality segmentation"""
    in_channels = 3
    out_channels = 1
    def __init__(
        self,
        images_dir,
        transform = None,
        image_size = 256,
        random_sampling = True,
    ):
        volumes = {}
        masks = {}
        print("reading images...")
        for (dirpath, dirnames, filenames) in os.walk(images_dir):
            image_slices = []
            mask_slices = []
            for filename in sorted(
                filter(lambda f: ".tif" in f, filenames),
                key=lambda x: int(x.split(".")[-2].split("_")[4]),
            ):
                filepath = os.path.join(dirpath, filename)
                if "mask" in filename:
                    mask_slices.append(imread(filepath, as_gray=True))
                else:
                    image_slices.append(imread(filepath))
            if len(image_slices) > 0:
                patient_id = dirpath.split("/")[-1]
                volumes[patient_id] = np.array(image_slices[1:-1])
                masks[patient_id] = np.array(mask_slices[1:-1])
        self.patients = sorted(volumes)
        print("preprocessing volumes...")
        # create list of tuples (volume, mask)
        self.volumes = [(volumes[k], masks[k]) for k in self.patients]
        print("cropping volumes...")
        # crop to smallest enclosing volume
        self.volumes = [crop_sample(v) for v in self.volumes]
        print("padding volumes...")
        # pad to square
        self.volumes = [pad_sample(v) for v in self.volumes]
        print("resizing volumes...")
        # resize
        self.volumes = [resize_sample(v, size=image_size) for v in self.volumes]
        print("normalizing volumes...")
        # normalize channel-wise
        self.volumes = [(normalize_volume(v), m) for v, m in self.volumes]
        # probabilities for sampling slices based on masks
        self.slice_weights = [m.sum(axis=-1).sum(axis=-1) for v, m in self.volumes]
        self.slice_weights = [
            (s + (s.sum() * 0.1 / len(s))) / (s.sum() * 1.1) for s in self.slice_weights
        ]
        # add channel dimension to masks
        self.volumes = [(v, m[..., np.newaxis]) for (v, m) in self.volumes]
        print("done creating dataset")
        # create global index for patient and slice (idx -> (p_idx, s_idx))
        num_slices = [v.shape[0] for v, m in self.volumes]
        self.patient_slice_index = list(
            zip(
                sum([[i] * num_slices[i] for i in range(len(num_slices))], []),
                sum([list(range(x)) for x in num_slices], []),
            )
        )
        self.random_sampling = random_sampling
        self.transform = transform
    def __len__(self):
        return len(self.patient_slice_index)
    def __getitem__(self, idx):
        patient = self.patient_slice_index[idx][0]
        slice_n = self.patient_slice_index[idx][1]
        if self.random_sampling:
            patient = np.random.randint(len(self.volumes))
            slice_n = np.random.choice(
                range(self.volumes[patient][0].shape[0]), p=self.slice_weights[patient]
            )
        v, m = self.volumes[patient]
        image = v[slice_n]
        mask = m[slice_n]
        if self.transform is not None:
            image, mask = self.transform((image, mask))
        # fix dimensions (C, H, W)
        image = image.transpose(2, 0, 1)
        mask = mask.transpose(2, 0, 1)
        image_tensor = torch.from_numpy(image.astype(np.float32))
        mask_tensor = torch.from_numpy(mask.astype(np.float32))
        # return tensors
        return image_tensor, mask_tensor

image_size = 256
aug_scale = 0.05
aug_angle = 15
flip_prob = 0.5

train_ds = BrainSegmentationDataset(
        images_dir = train_dir,
        image_size = image_size,
        transform = transforms(scale = aug_scale, angle = aug_angle, flip_prob=0.5),
        random_sampling = True
)

valid_ds = BrainSegmentationDataset(
        images_dir = valid_dir,
        image_size = image_size,
        random_sampling=False
)

batch_size = 4

train_loader = torch.utils.data.DataLoader(
        train_ds,
        batch_size = batch_size,
        shuffle = True,
        drop_last = True,
        num_workers = 8
)

valid_loader = torch.utils.data.DataLoader(
        valid_ds,
        batch_size = batch_size,
        drop_last = False,
)

dataloaders = {"train": train_loader, "valid": valid_loader}
image, mask = next(iter(train_loader))
image.size()
mask.size()

dataset_sizes = {x: len(dataloaders[x]) for x in ['train', 'valid']}
print(dataset_sizes)

class UNet(nn.Module):
    def __init__(
        self,
        channels_in = 3,
        n_classes = 1,
        depth = 5,
        n_filters = 6, 
    ):
        super(UNet, self).__init__()
        self.depth = depth
        prev_channels = channels_in
        self.down_path = nn.ModuleList()
        for i in range(depth):
            self.down_path.append(
                DownBlock(prev_channels, 2 ** (n_filters + i))
            )
            prev_channels = 2 ** (n_filters + i)
        self.up_path = nn.ModuleList()
        for i in reversed(range(depth - 1)):
            self.up_path.append(
                UpBlock(prev_channels, 2 ** (n_filters + i))
            )
            prev_channels = 2 ** (n_filters + i)
        self.last = nn.Conv2d(prev_channels, n_classes, kernel_size = 1)
    def forward(self, x):
        blocks = []
        for i, down in enumerate(self.down_path):
            x = down(x)
            if i != len(self.down_path) - 1:
                blocks.append(x)
                x = F.max_pool2d(x, 2)
                #print("after maxpool: x is {}".format(x.size()))
        for i, up in enumerate(self.up_path):
            x = up(x, blocks[-i - 1])
        return torch.sigmoid(self.last(x))

class ConvBlock(nn.Module):
    def __init__(self, in_size, out_size):
        super(ConvBlock, self).__init__()
        block = []
        block.append(nn.Conv2d(in_size, out_size, kernel_size = 3, padding = 1))
        block.append(nn.ReLU())
        #block.append(nn.BatchNorm2d(out_size))
        block.append(nn.Dropout(0.6))
        block.append(nn.Conv2d(out_size, out_size, kernel_size = 3, padding = 1))
        block.append(nn.ReLU())
        #block.append(nn.BatchNorm2d(out_size))
        block.append(nn.Dropout(0.6))
        self.block = nn.Sequential(*block)
    def forward(self, x):
        out = self.block(x)
        return out

class UpBlock(nn.Module):
    def __init__(self, in_size, out_size):
        super(UpBlock, self).__init__()
        self.up = nn.ConvTranspose2d(in_size, out_size, kernel_size = 2, stride = 2)
        self.conv_block = ConvBlock(in_size, out_size)
    def forward(self, x, bridge):
        up = self.up(x)
        #print("in upblock forward: up is {}".format(up.size()))
        #print("in upblock forward: bridge is {}".format(bridge.size()))
        out = torch.cat([up, bridge], 1)
        #print("in upblock forward: concatenated is {}".format(out.size()))
        out = self.conv_block(out)
        #print("in upblock forward: out is {}".format(out.size()))
        return out

class DownBlock(nn.Module):
    def __init__(self, in_size, out_size):
        super(DownBlock, self).__init__()
        self.conv_block = ConvBlock(in_size, out_size)
    def forward(self, x):
        down = self.conv_block(x)
        #print("in downblock forward: down is {}".format(down.size()))
        return down



device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
model = UNet(depth = 5).to(device)


class DiceLoss(nn.Module):
    def __init__(self):
        super(DiceLoss, self).__init__()
        self.smooth = 1.0
    def forward(self, y_pred, y_true):
        assert y_pred.size() == y_true.size()
        y_pred = y_pred[:, 0].contiguous().view(-1)
        y_true = y_true[:, 0].contiguous().view(-1)
        intersection = (y_pred * y_true).sum()
        dsc = (2. * intersection + self.smooth) / (
            y_pred.sum() + y_true.sum() + self.smooth
        )
        return 1. - dsc


dice_loss = DiceLoss()
bce_loss = nn.BCELoss()
dice_weight = 0.3

optimizer = torch.optim.SGD(model.parameters(), lr = 0.1, momentum = 0.9)

scheduler = torch.optim.lr_scheduler.OneCycleLR(
    optimizer,
    max_lr = 0.1,
    steps_per_epoch = len(train_loader),
    epochs = num_epochs
)

best_model_wts = copy.deepcopy(model.state_dict())
best_dice_coef = 0.0


num_epochs = 50

for epoch in range(num_epochs):
    print('Epoch {}/{}'.format(epoch, num_epochs - 1), flush = True)
    print('-' * 10)
    for phase in ['train', 'valid']:
        print("Entering phase: " + phase, flush = True)
        if phase == 'train':
            model = model.train() 
        else:
            model = model.eval()   
        running_loss = 0.0
        running_dice = 0.0
        running_bce = 0.0
        for inputs, labels in dataloaders[phase]:
            inputs = inputs.to(device)
            labels = labels.to(device)
            optimizer.zero_grad()
            with torch.set_grad_enabled(phase == 'train'):
                preds = model(inputs)
                dice_loss = dsc_loss(preds, labels)
                xent_loss = bce_loss(preds, labels)
                loss = dice_weight * dice_loss + (1 - dice_weight) * xent_loss
                if phase == 'train':
                    loss.backward()
                    optimizer.step()
            running_loss += loss.item() * inputs.size(0)
            running_dice += dice_loss.item() * inputs.size(0)
            running_bce += xent_loss.item() * inputs.size(0)
            # The 1cycle learning rate policy changes the learning rate after every batch. 
            # step should be called after a batch has been used for training.
            if phase == 'train':
                scheduler.step()
              #print(scheduler.get_last_lr(), flush = True)
        # these are per batch!
        epoch_loss = running_loss / dataset_sizes[phase]
        epoch_dice = running_dice / dataset_sizes[phase]
        epoch_bce = running_bce / dataset_sizes[phase]
        if phase == 'valid' and epoch_dice < best_dice_coef:
            best_dice = epoch_dice
            best_model_wts = copy.deepcopy(model.state_dict())
        print('{} Loss: {:.4f}'.format(phase, epoch_loss), flush = True)
        print('{} Dice coef: {:.4f}'.format(phase, epoch_dice), flush = True)
        print('{} BCE: {:.4f}'.format(phase, epoch_bce), flush = True)
    print()

model.load_state_dict(best_model_wts)
torch.save(model.state_dict(), "mri.pt")

File: scripts/unet.R

library(torch)
library(torchvision)
library(tidyverse)
library(magick)
library(zeallot)
library(cowplot)
library(zip)
library(pins)


# Get data ----------------------------------------------------------------
torch_manual_seed(777)
set.seed(777)

pins::board_register_kaggle(token = "~/kaggle.json")

files <- pins::pin_get("mateuszbuda/lgg-mri-segmentation", board = "kaggle",  extract = FALSE)

train_dir <- "data/mri_train"
valid_dir <- "data/mri_valid"

if(dir.exists(train_dir)) unlink(train_dir, recursive = TRUE, force = TRUE)
if(dir.exists(valid_dir)) unlink(valid_dir, recursive = TRUE, force = TRUE)

zip::unzip(files, exdir = "data")

file.rename("data/kaggle_3m", train_dir)

# this is a duplicate
unlink("data/lgg-mri-segmentation", recursive = TRUE)

dir.create(valid_dir)

patients <- list.dirs(train_dir, recursive = FALSE)
length(patients)

valid_indices <- sample(1:length(patients), 30)
valid_indices

patients[valid_indices]

for (i in valid_indices) {
  dir.create(file.path(valid_dir, basename(patients[i])))
  for (f in list.files(patients[i])) {    
    file.rename(file.path(train_dir, basename(patients[i]), f), file.path(valid_dir, basename(patients[i]), f))    
  }
  unlink(file.path(train_dir, basename(patients[i])), recursive = TRUE)
}

list.dirs(train_dir, recursive = FALSE) %>% length()

# Dataset -----------------------------------------------------------------

brainseg_dataset <- dataset(
  name = "brainseg_dataset",
  
  initialize = function(img_dir,
                        augmentation_params = NULL,
                        random_sampling = FALSE) {
    self$images <- tibble(
      img = grep(
        list.files(
          img_dir,
          full.names = TRUE,
          pattern = "tif",
          recursive = TRUE
        ),
        pattern = 'mask',
        invert = TRUE,
        value = TRUE
      ),
      mask = grep(
        list.files(
          img_dir,
          full.names = TRUE,
          pattern = "tif",
          recursive = TRUE
        ),
        pattern = 'mask',
        value = TRUE
      )
    )
    self$slice_weights <- self$calc_slice_weights(self$images$mask)
    self$augmentation_params <- augmentation_params
    self$random_sampling <- random_sampling
  },
  
  .getitem = function(i) {
    index <-
      if (self$random_sampling == TRUE)
        sample(1:self$.length(), 1, prob = self$slice_weights)
    else
      i
    
    img <- self$images$img[index] %>%
      image_read() %>%
      transform_to_tensor() 
    mask <- self$images$mask[index] %>%
      image_read() %>%
      transform_to_tensor() %>%
      transform_rgb_to_grayscale() %>%
      torch_unsqueeze(1)
    
    img <- self$min_max_scale(img)
    
    if (!is.null(self$augmentation_params)) {
      scale_param <- self$augmentation_params[1]
      c(img, mask) %<-% self$resize(img, mask, scale_param)
      
      rot_param <- self$augmentation_params[2]
      c(img, mask) %<-% self$rotate(img, mask, rot_param)
      
      flip_param <- self$augmentation_params[3]
      c(img, mask) %<-% self$flip(img, mask, flip_param)
      
    }
    list(img = img, mask = mask)
  },
  
  .length = function() {
    nrow(self$images)
  },
  
  calc_slice_weights = function(masks) {
    weights <- map_dbl(masks, function(m) {
      img <-
        as.integer(magick::image_data(image_read(m), channels = "gray"))
      sum(img / 255)
    })
    
    sum_weights <- sum(weights)
    num_weights <- length(weights)
    
    weights <- weights %>% map_dbl(function(w) {
      w <- (w + sum_weights * 0.1 / num_weights) / (sum_weights * 1.1)
    })
    weights
  },
  
  min_max_scale = function(x) {
    min = x$min()$item()
    max = x$max()$item()
    x$clamp_(min = min, max = max)
    x$add_(-min)$div_(max - min + 1e-5)
    x
  },
  
  resize = function(img, mask, scale_param) {
    img_size <- dim(img)[2]
    rnd_scale <- runif(1, 1 - scale_param, 1 + scale_param)
    img <- transform_resize(img, size = rnd_scale * img_size)
    mask <- transform_resize(mask, size = rnd_scale * img_size)
    diff <- dim(img)[2] - img_size
    if (diff > 0) {
      top <- ceiling(diff / 2)
      left <- ceiling(diff / 2)
      img <- transform_crop(img, top, left, img_size, img_size)
      mask <- transform_crop(mask, top, left, img_size, img_size)
    } else {
      img <- transform_pad(img,
                           padding = -c(
                             ceiling(diff / 2),
                             floor(diff / 2),
                             ceiling(diff / 2),
                             floor(diff / 2)
                           ))
      mask <- transform_pad(mask, padding = -c(
        ceiling(diff / 2),
        floor(diff /
                2),
        ceiling(diff /
                  2),
        floor(diff /
                2)
      ))
    }
    list(img, mask)
  },
  
  rotate = function(img, mask, rot_param) {
    rnd_rot <- runif(1, 1 - rot_param, 1 + rot_param)
    img <- transform_rotate(img, angle = rnd_rot)
    mask <- transform_rotate(mask, angle = rnd_rot)
    
    list(img, mask)
  },
  
  flip = function(img, mask, flip_param) {
    rnd_flip <- runif(1)
    if (rnd_flip > flip_param) {
      img <- transform_hflip(img)
      mask <- transform_hflip(mask)
    }
    
    list(img, mask)
  }
)

train_ds <-
  brainseg_dataset(
    train_dir,
    augmentation_params = c(0.05, 15, 0.5),
    random_sampling = TRUE
  )

length(train_ds)

valid_ds <-
  brainseg_dataset(valid_dir,
                   augmentation_params = NULL,
                   random_sampling = FALSE)

length(valid_ds)

par(mfrow = c(1, 2), mar = c(0, 1, 0, 1))
img_and_mask <- valid_ds[27]
img <- img_and_mask[[1]]
mask <- img_and_mask[[2]]
img$permute(c(2, 3, 1)) %>% as.array() %>% as.raster() %>% plot()
mask$squeeze() %>% as.array() %>% as.raster() %>% plot()



# Plot augmentation -------------------------------------------------------

img_and_mask <- valid_ds[77]
img <- img_and_mask[[1]]
mask <- img_and_mask[[2]]

imgs <- map (1:24, function(i) {
  c(img, mask) %<-% valid_ds$resize(img, mask, 0.2) # train_ds really uses 0.05
  c(img, mask) %<-% valid_ds$flip(img, mask, 0.5)
  c(img, mask) %<-% valid_ds$rotate(img, mask, 90) # train_ds really uses 15
  img %>%
    transform_rgb_to_grayscale() %>%
    as.array() %>%
    as_tibble() %>%
    rowid_to_column(var = "Y") %>%
    gather(key = "X", value = "value", -Y) %>%
    mutate(X = as.numeric(gsub("V", "", X))) %>%
    ggplot(aes(X, Y, fill = value)) +
    geom_raster() +
    theme_void() +
    theme(legend.position = "none") +
    theme(aspect.ratio = 1)
})

plot_grid(plotlist = imgs, nrow = 4)



# Load data ---------------------------------------------------------------

batch_size <- 4
train_dl <- dataloader(train_ds, batch_size)
valid_dl <- dataloader(valid_ds, batch_size)



# Model -------------------------------------------------------------------

unet <- nn_module(
  "unet",
  
  initialize = function(channels_in = 3,
                        n_classes = 1,
                        depth = 5,
                        n_filters = 6) {
    
    self$down_path <- nn_module_list()
    
    prev_channels <- channels_in
    for (i in 1:depth) {
      self$down_path$append(down_block(prev_channels, 2 ^ (n_filters + i - 1)))
      prev_channels <- 2 ^ (n_filters + i -1)
    }
    
    self$up_path <- nn_module_list()
    for (i in ((depth - 1):1)) {
      self$up_path$append(up_block(prev_channels, 2 ^ (n_filters + i - 1)))
      prev_channels <- 2 ^ (n_filters + i - 1)
    }
    
    self$last = nn_conv2d(prev_channels, n_classes, kernel_size = 1)
  },
  
  forward = function(x) {
    
    blocks <- list()
    
    for (i in 1:length(self$down_path)) {
      x <- self$down_path[[i]](x)
      if (i != length(self$down_path)) {
        blocks <- c(blocks, x)
        x <- nnf_max_pool2d(x, 2)
      }
    }
    
    for (i in 1:length(self$up_path)) {  
      x <- self$up_path[[i]](x, blocks[[length(blocks) - i + 1]]$to(device = device))
    }
    torch_sigmoid(self$last(x))
  }
)

down_block <- nn_module(
  "down_block",
  
  initialize = function(in_size, out_size) {
    self$conv_block <- conv_block(in_size, out_size)
  },
  
  forward = function(x) {
    self$conv_block(x)
  }
)

up_block <- nn_module(
  "up_block",
  initialize = function(in_size, out_size) {
    self$up = nn_conv_transpose2d(in_size,
                                  out_size,
                                  kernel_size = 2,
                                  stride = 2)
    self$conv_block = conv_block(in_size, out_size)
    
  } ,
  forward = function(x, bridge) {
    up <- self$up(x)
    torch_cat(list(up, bridge), 2) %>%
      self$conv_block()
  }
)

conv_block <- nn_module( 
  "conv_block",
  
  initialize = function(in_size, out_size) {
    
    self$conv_block <- nn_sequential(
      nn_conv2d(in_size, out_size, kernel_size = 3, padding = 1),
      nn_relu(),
      nn_dropout(0.6),
      nn_conv2d(out_size, out_size, kernel_size = 3, padding = 1),
      nn_relu()
    )
  },
  
  forward = function(x){
    self$conv_block(x)
  }
)


device <- torch_device(if(cuda_is_available()) "cuda" else "cpu")
model <- unet(depth = 5)$to(device = device)


# Loss, optimizer etc -----------------------------------------------------

calc_dice_loss <- function(y_pred, y_true) {
  smooth <- 1
  y_pred <- y_pred$view(-1)
  y_true <- y_true$view(-1)
  intersection <- (y_pred * y_true)$sum()
  1 - ((2 * intersection + smooth) / (y_pred$sum() + y_true$sum() + smooth))
}

dice_weight <- 0.3

optimizer <- optim_sgd(model$parameters, lr = 0.1, momentum = 0.9)

num_epochs <- 20
scheduler <- lr_one_cycle(optimizer, max_lr = 0.1, steps_per_epoch = length(train_dl), epochs = num_epochs)


# Train -------------------------------------------------------------------

train_batch <- function(b) {
  
  optimizer$zero_grad()
  output <- model(b[[1]]$to(device = device))
  target <- b[[2]]$to(device = device)
  
  bce_loss <- nnf_binary_cross_entropy(output, target)
  dice_loss <- calc_dice_loss(output, target)
  loss <-  dice_weight * dice_loss + (1 - dice_weight) * bce_loss
  
  loss$backward()
  optimizer$step()
  scheduler$step()
  
  gc(full = TRUE)
  list(bce_loss$item(), dice_loss$item(), loss$item())
  
}

valid_batch <- function(b) {
  
  output <- model(b[[1]]$to(device = device))
  target <- b[[2]]$to(device = device)
  
  if (i %% 77 == 0) {
    first <- output[1, , , ]
    inferred_mask <- first$to(device = "cpu") %>% as.array() %>% .[1, ,]
    print(quantile(inferred_mask))
  }
  
  bce_loss <- nnf_binary_cross_entropy(output, target)
  dice_loss <- calc_dice_loss(output, target)
  loss <-  dice_weight * dice_loss + (1 - dice_weight) * bce_loss
  
  gc(full = TRUE)
  list(bce_loss$item(), dice_loss$item(), loss$item())
}

for (epoch in 1:num_epochs) {
  
  model$train()
  train_bce <- c()
  train_dice <- c()
  train_loss <- c()
  
  
  for (b in enumerate(train_dl)) {
    c(bce_loss, dice_loss, loss) %<-% train_batch(b)
    train_bce <- c(train_bce, bce_loss)
    train_dice <- c(train_dice, dice_loss)
    train_loss <- c(train_loss, loss)
  }
  
  torch_save(model, paste0("model_", epoch, ".pt"))
  cat(sprintf("\nEpoch %d, training: loss:%3f, bce: %3f, dice: %3f\n", epoch, mean(train_loss), mean(train_bce), mean(train_dice)))
  
  model$eval()
  valid_bce <- c()
  valid_dice <- c()
  valid_loss <- c()
  
  i <- 0
  for (b in enumerate(valid_dl)) {
    
    i <<- i + 1
    c(bce_loss, dice_loss, loss) %<-% valid_batch(b)
    valid_bce <- c(valid_bce, bce_loss)
    valid_dice <- c(valid_dice, dice_loss)
    valid_loss <- c(valid_loss, loss)
    
  }
  
  cat(sprintf("\nEpoch %d, validation: loss:%3f, bce: %3f, dice: %3f\n", epoch, mean(valid_loss), mean(valid_bce), mean(valid_dice)))
}

saved_model <- torch_load("model_20.pt") 
model <- saved_model
model$eval()

eval_ds <- brainseg_dataset(valid_dir,
                            augmentation_params = NULL,
                            random_sampling = TRUE)
eval_dl <- dataloader(eval_ds, batch_size = 8)

batch <- eval_dl %>% dataloader_make_iter() %>% dataloader_next()

par(mfcol = c(3, 8), mar = c(0, 1, 0, 1))

for (i in 1:8) {
  
  img <- batch[[1]][i, .., drop = FALSE]
  inferred_mask <- model(img$to(device = device))
  true_mask <- batch[[2]][i, ..,drop = FALSE]$to(device = device)
  
  bce <- nnf_binary_cross_entropy(inferred_mask, true_mask)$to(device = "cpu") %>% as.numeric()
  dc <- calc_dice_loss(inferred_mask, true_mask)$to(device = "cpu") %>% as.numeric()
  cat(sprintf("\nSample %d, bce: %3f, dice: %3f\n", i, bce, dc))
  
  
  inferred_mask <- inferred_mask$to(device = "cpu") %>% as.array() %>% .[1, 1, , ]
  
  inferred_mask <- ifelse(inferred_mask > 0.5, 1, 0)
  
  img[1, 1, , ] %>% as.array() %>% as.raster() %>% plot()
  true_mask$to(device = "cpu")[1, 1, , ] %>% as.array() %>% as.raster() %>% plot()
  inferred_mask %>% as.raster() %>% plot()
}

